### ISSUE TITLE
Feature: plot n layers' weights every m epochs

### ISSUE THREAD

**Issue by EmbraceLife** [EmbraceLife/LIE#3]
secondary title: Why Kur over TF, Keras, Pytorch, ConvNet.js?
- Easy debugging with logger setup 
- Deep Learning Infrastructure or workflow is fully covered and ready to use with little customization 
- Given the above two, it's easy to learn what is going on inside by experimenting
- Javascript with DL makes easy to visualize training process, but in theory any DL lib could allow you to visualize the workflow step by step. The easier to hack, the more you can do what you want.
- A challenge (mnist): can I visualize every step from an image, to a  processed data sample, to first layer weights and output, to second ..., logits and loss, ..., second layer weights, first layer weights
- The key: find the detail working of data from in and out of a layer, add a plot func ...?

**Comment by EmbraceLife** [comment]
How to access weights? 
- weight folders saving weights, e.g., best.valid.w folder update the training weights with best validation loss 
- weights can be saved and loaded with `utils/idx.py`, back and forth between format np.array and idx

How to plot weights? 
- borrow [functions](https://hyp.is/MKzd7C4eEeeWlPvso_EWdg/nbviewer.jupyter.org/github/Hvass-Labs/TensorFlow-Tutorials/blob/master/01_Simple_Linear_Model.ipynb) from Hvass tutorials 
- plot pure weights 
- plot an image applied to weights? No! applying an image array is to find the right label using the help of weights

How to visualize the changes of weights as loss is dropping? 
- plot the weight in the first epoch
- plot weights every 100 epoch
- save loss values with each plotting
- these actions' trigger are tested after each epoch 
- Now, we can save all weights plotting every 100 epochs with this [commit](https://github.com/EmbraceLife/LIE/commit/79f8cf7e21f5cc08494bb6ea98b24eaf665e2b9e) looking for `plot_weights` in `executor.py`
- finally get this `plot_weights()` into a training_hook during `wrapped_train`

**Comment by EmbraceLife** [comment]
By now [commit](https://github.com/EmbraceLife/LIE/commit/1f90a0368245d2bde588aba50fc66d455d8620c3), I have implemented a hook for plotting weights: `plot_weights_hook.py` in `kur/model/hooks` folder
- run `kur train mnist1.yml`, every 100 epochs, a figure on weights will be created into the folder of `plot_weights/`
- however, this hook is plotting only for `mnist1.yml` at this moment, it needs to plot for all weights rather than just a particular weight file. See inside `plot_weights_hooks.py`: 
```
def plot_weights():
			w = idx.load("mnist.best.valid.w/..dense.0+..dense.0_kernel:0.kur")
```

**Comment by EmbraceLife** [comment]
I tried to automatically identify the filename for kernels, but there are more technical details to deal with before it can print all kernels automatically. 

to keep it simple, I shall choose a kernel to print!!! it is more flexible. Also different kernels have different shapes, especially for convolutional hidden layers ...

Here is a conflict at line 355 of `executor.py`: 
```
# ?????? to comment back: conflict with info['epoch']
			if training_hooks:
				for hook in training_hooks:
					hook.notify(
						TrainingHook.TRAINING_END,
						log=log,
						info={'Reason' : reason}
					)
```

**Comment by EmbraceLife** [comment]
How I want to use plot_weights hook: 

```yaml
hooks: 
  - plot_weights: # this is working
      filename: weight_filename # don't know how to write this up in kur????
```

A lots can be learnt from `plot_hook.py`, I think

**Comment by EmbraceLife** [comment]
**How to plot one or more layers' weights?**
Now, I can plot the convolutional weights after 10 epochs of training: 
```yaml
train:
  hooks:
    - plot_weights:
        plot_every_n_epochs: 10
        weight_file: cifar.best.valid.w
        weight_keywords1: ["convolution.0", "kernel"]
        weight_keywords2: ["convolution.1", "kernel"]
```

using `plot_weights_hook.py` to set more `weight_keywords3`, `weight_keyword4` if needed

```python
def __init__(self, weight_file, weight_keywords1, weight_keywords2, weight_keywords3, *args, **kwargs):
		""" Creates a new plotting hook, get plot filenames and matplotlib ready.
		"""
```

my question: 
if I want to plot more convolutional weights, say `weight_keywords3`, `weight_keywords4`, `weight_keywords5`, do I have to change the source code, by adding them into `__init__` above? 

Can `**kwargs` somehow help me here? If so, how?

**Comment by EmbraceLife** [comment]
**How to enable this feature in standard kur?**
- paste two files 
- `kur/model/hooks/__init__.py`
- `kur/model/hooks/plot_weights_hook.py`
- Then it should be ready to use this feature
